### PR TITLE
agent service: new protocol & namespaces

### PR DESCRIPTION
**Pull request by theomonnom**
This PR depends on this [protocol improvements PR](https://github.com/livekit/protocol/pull/595)

My Go may be rusty, so don't hesitate to make me redo nit things to make stuff idiomatic 😊

### CODE REVIEW

File: pkg/agent/worker.go [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+	}()
+
+	return w
+}
+
+func (w *Worker) sendRequest(req *livekit.ServerMessage) {
+	w.msgChan <- req
+}
+
+func (w *Worker) ReadChan() <-chan *livekit.ServerMessage {
```
Comment (theomonnom) [inline pkg/agent/worker.go:103]: Not sure how idiomatic it is to create this `ReadChan` method.
The motivation is that the WSSignalConnection is inside the service pkg, to avoid cyclic dependencies, the worker can't have this object in its field. So I'm communicating with the agentservice using this channel

File: pkg/agent/client.go [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+		nsMap := make(map[string]bool)
+
+		for r := range resChan {
+			if r.Result.GetRoomEnabled() && !roomEnabled.Resolved() {
+				roomEnabled.Resolve(true, nil)
+			}
+			if r.Result.GetPublisherEnabled() && !publisherEnabled.Resolved() {
+				publisherEnabled.Resolve(true, nil)
+			}
+			if r.Result.GetNamespaces() != nil {
```
Comment (theomonnom) [inline pkg/agent/client.go:165]: Wondering how slow it is to fetch the used namespaces from all servers..

File: pkg/rtc/room.go [hunk 3]
```diff
@@ -180,21 +180,6 @@ func NewRoom(
 		r.protoRoom.CreationTime = time.Now().Unix()
 	}
 
-	if agentClient != nil {
```
Comment (theomonnom) [inline pkg/rtc/room.go:183]: See https://live-kit.slack.com/archives/C02L4JW8ACT/p1709648118669889

File: pkg/service/agentservice.go [hunk 4]
```diff
@@ -1,17 +1,3 @@
-// Copyright 2023 LiveKit, Inc.
```
Comment (davidzhao) [inline pkg/service/agentservice.go:1]: should not remove copyright notice

File: pkg/service/agentservice.go [hunk 5]
```diff
@@ -116,64 +103,55 @@ func (s *AgentService) ServeHTTP(writer http.ResponseWriter, r *http.Request) {
...
-		agentServer:      agentServer,
-		roomTopic:        roomTopic,
-		publisherTopic:   publisherTopic,
-		availability:     make(map[string]chan *availability),
-		unregistered:     make(map[*websocket.Conn]*worker),
-		roomWorkers:      make(map[string]*worker),
-		publisherWorkers: make(map[string]*worker),
+func (s *AgentService) HandleConnection(r *http.Request, conn *websocket.Conn) {
+	var protocol agent.WorkerProtocolVersion
+	if pv, err := strconv.Atoi(r.Form.Get("protocol")); err == nil {
```
Comment (davidzhao) [inline pkg/service/agentservice.go:111]: `request.Form` is not populated unless explicitly parsed. better to use [FormValue](https://pkg.go.dev/net/http#Request.FormValue)

File: pkg/service/agentservice.go [hunk 6]
```diff
@@ -188,321 +166,207 @@ func (s *AgentHandler) HandleConnection(ctx context.Context, conn *websocket.Con
...
-		} else if !s.roomRegistered && s.roomAvailableLocked() {
-			if err := s.agentServer.RegisterJobRequestTopic(s.roomTopic); err != nil {
-				w.logger.Errorw("failed to register room agents", err)
-			} else {
-				s.roomRegistered = true
-			}
+	if worker.JobType() == livekit.JobType_JT_PUBLISHER {
+		info.numPublishers--
+		if info.numPublishers == 0 {
+			h.agentServer.DeregisterJobRequestTopic(worker.Namespace(), agent.PublisherAgentTopic)
```
Comment (davidzhao) [inline pkg/service/agentservice.go:249]: I guess the change in topics means all of the server instances would have to be updated at the same time. good to keep in mind.

File: pkg/service/agentservice.go [hunk 7]
```diff
@@ -188,321 +166,207 @@ func (s *AgentHandler) HandleConnection(ctx context.Context, conn *websocket.Con
...
-						selected.logger.Errorw("failed to assign job", err, "workerID", selected.id)
-					} else {
-						selected.mu.Lock()
-						selected.activeJobs++
-						selected.mu.Unlock()
-						return &emptypb.Empty{}, nil
-					}
-				}
+		if w.Status() == livekit.WorkerStatus_WS_AVAILABLE {
+			if len(w.RunningJobs()) > 0 {
```
Comment (davidzhao) [inline pkg/service/agentservice.go:299]: not part of this PR, but this could be more sophisticated.

File: pkg/service/roomservice.go [hunk 8]
```diff
@@ -102,10 +102,11 @@ func (s *RoomService) CreateRoom(ctx context.Context, req *livekit.CreateRoomReq
 
 	if created {
 		go func() {
-			s.agentClient.JobRequest(ctx, &livekit.Job{
-				Id:   utils.NewGuid("JR_"),
-				Type: livekit.JobType_JT_ROOM,
-				Room: rm,
+			res := s.agentClient.CheckEnabled(ctx, &rpc.CheckEnabledRequest{})
+			s.agentClient.JobRequest(ctx, &agent.JobDescription{
```
Comment (davidzhao) [inline pkg/service/roomservice.go:106]: shouldn't it check if it's enabled before firing this off?

File: test/agent_test.go [hunk 9]
```diff
@@ -117,6 +117,36 @@ func TestAgentNamespaces(t *testing.T) {
...
+	require.Equal(t, int32(1), ac2.registered.Load())
+
+	c1 := createRTCClient("c1", secondServerPort, nil) // Create a room on the second node
+	waitUntilConnected(t, c1)
+
+	t1, err := c1.AddStaticTrack("audio/opus", "audio", "micro")
+	require.NoError(t, err)
+	defer t1.Stop()
+
+	time.Sleep(time.Second * 10)
```
[the remark this answers is not in the document]
Comment (davidzhao) [reply test/agent_test.go:144]: that's not expected.. but perhaps it's due to the way we are waiting for namespaces to load.

another thing in tests, it's better to use [WithTimeout](https://github.com/livekit/livekit/blob/d87f8aa299a3d9b27dde14fcb2d586683b431084/pkg/testutils/timeout.go#L27) rather than hard sleeps.

**Review by davidzhao — COMMENTED** [review]
nice work! mostly minor comments

File: pkg/agent/job.go [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+		if j.status == livekit.JobStatus_JS_FAILED {
+			j.Logger.Errorw("job failed", nil, "id", j.id, "type", j.jobType, "error", req.Error)
+		}
+	}
+
+	j.load = req.Load
+	j.mu.Unlock()
+
+	if req.Metadata != nil {
+		j.UpdateMetadata(req.GetMetadata())
```
[the remark this answers is not in the document]
Comment (theomonnom) [reply pkg/agent/job.go:66]: Metadata could be useful when we implement ListJobs.
Also I'm thinking about showing the metadata changes inside the dashboard so people can debug the job timelime

File: pkg/agent/worker.go (hunk 1 above)
Comment (theomonnom) [reply pkg/agent/worker.go:103]: Not sure to understand, should I use routing.MessageSink?

**Comment by theomonnom** [comment]
Backward compatibility is working. Tested with the main branch of agents

File: pkg/utils/incrementaldispatcher.go [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+		d.lock.RUnlock()
+		for _, item := range itemsToDispatch {
+			fn(item)
+		}
+	}
+	for !d.done.IsBroken() {
+		dispatchFromIdx()
+		d.lock.Lock()
+		if idx == len(d.items) {
+			d.cond.Wait()
```
Comment (theomonnom) [inline pkg/utils/incrementaldispatcher.go:72]: No need to unlock before waiting?

Comment (davidzhao) [reply pkg/utils/incrementaldispatcher.go:72]: nope.. it Cond unlocks while it's waiting.. and locks before returning control

**Review by frostbyte73 — APPROVED** [review]

**Review by davidzhao — APPROVED** [review]

File: pkg/service/rtcservice.go [hunk 12]
```diff
@@ -541,16 +548,6 @@ func (s *RTCService) startConnection(
...
-	if created && s.agentClient != nil {
-		go func() {
-			s.agentClient.JobRequest(ctx, &livekit.Job{
-				Id:   putil.NewGuid("JR_"),
-				Type: livekit.JobType_JT_ROOM,
-				Room: cr.Room,
-			})
-		}()
-	}
-
```
Comment (theomonnom) [inline pkg/service/rtcservice.go:553]: cc @davidzhao @keepingitneil
